http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala

```diff
@@ -194,6 +194,32 @@ class SamAuthProvider[F[_]: OpenTelemetryMetrics](
     } yield roles.contains(SamRole.Owner)
   }
 
+  def filterWorkspaceReader(
+    resources: NonEmptyList[WorkspaceResourceSamResourceId],
+    userInfo: UserInfo
+  )(implicit
+    ev: Ask[F, TraceId]
+  ): F[Set[WorkspaceResourceSamResourceId]] = {
+    val authHeader = Authorization(Credentials.Token(AuthScheme.Bearer, userInfo.accessToken.token))
+    for {
+      workspacePolicies <- samDao
+        .getResourcePolicies[WorkspaceResourceSamResourceId](authHeader, SamResourceType.Workspace)
+      readingWorkspaces = workspacePolicies.collect { case (r, p) =>
+        r
+      }
+    } yield readingWorkspaces.toSet
+  }
+
+  def isUserWorkspaceReader(
+    workspaceResource: WorkspaceResourceSamResourceId,
+    userInfo: UserInfo
+  )(implicit ev: Ask[F, TraceId]): F[Boolean] = {
+    val authHeader = Authorization(Credentials.Token(AuthScheme.Bearer, userInfo.accessToken.token))
+    for {
+      roles <- samDao.getResourceRoles(authHeader, workspaceResource)
```

Comment:
Same thinking here, if there is a role associated with the user, it is at least reader

Reply:
You could also explicitly check the read permission by `samDao.hasResourcePermission(workspaceResource, SamAction.Read, authHeader)`

Reply:
That sounds like a cleaner way that does not rely on a side effect, I will implement that, thanks!

Reply:
I might actually stick to checking the user role, there is no clear mapping to  a read sam action that I can see in their doc (unless read_policies maybe?)